File: pattern_searching/kmp_pattern_searching.cc
#include <vector>

#include "utils/test.h"

namespace algorithm {
namespace pattern_searching {

template <std::size_t N, std::size_t M>
std::vector<int32_t> KMPPatternSearching(const char (&text)[N],
                                         const char (&pattern)[M]) {
  if (N < 2 || M < 2 || N < M) {
    return std::vector<int32_t>();
  }

  uint32_t lps[M];
  uint32_t i = 1, j = 0;
  std::vector<int32_t> ret;

  lps[0] = 0;
  while (i < M - 1) {
    if (pattern[i] == pattern[j]) {
      lps[i++] = ++j;
    } else if (j > 0) {
      j = lps[j - 1];
    } else {
      lps[i++] = 0;
    }
  }

  i = 0, j = 0;
  while (i < N - 1) {
    if (text[i] == pattern[j]) {
      i++;
      j++;
    }
    if (j == M - 1) {
      ret.push_back(i - j);
      j = lps[j - 1];
    } else if (i < N - 1 && pattern[j] != text[i]) {
      if (j != 0) {
        j = lps[j - 1];
      } else {
        i++;
      }
    }
  }

  return ret;
}

namespace test {

void KMPPatternSearching_HandlesMultiplePatternInput() {
  auto ret = KMPPatternSearching("AABAACAADAABAAABAA", "AABA");
  std::vector<int32_t> result{0, 9, 13};
  EXPECT_EQ(ret.size(), result.size());
  for (auto i = 0; i < ret.size(); i++) {
    EXPECT_EQ(ret[i], result[i]);
  }

  ret = KMPPatternSearching("AAABSSSAAAOSI", "FHJK");
  result.clear();
  EXPECT_EQ(ret.size(), result.size());
  for (auto i = 0; i < ret.size(); i++) {
    EXPECT_EQ(ret[i], result[i]);
  }

  ret = KMPPatternSearching("AAACAAIOAAADAAIS", "AAACA");
  result = {0};
  EXPECT_EQ(ret.size(), result.size());
  for (auto i = 0; i < ret.size(); i++) {
    EXPECT_EQ(ret[i], result[i]);
  }
}

void KMPPatternSearching_HandlesInvalidInput() {
  EXPECT_EQ(KMPPatternSearching("", "").size(), 0);
  EXPECT_EQ(KMPPatternSearching("ABC", "").size(), 0);
  EXPECT_EQ(KMPPatternSearching("", "DEF").size(), 0);
  EXPECT_EQ(KMPPatternSearching("AA", "DEF").size(), 0);
}

void RunTests() {
  TEST(KMPPatternSearching_HandlesMultiplePatternInput);
  TEST(KMPPatternSearching_HandlesInvalidInput);
}

}  // namespace test

}  // namespace pattern_searching
}  // namespace algorithm

int main() {
  std::cout << "-----Running KMP pattern searching tests-----" << std::endl;
  algorithm::pattern_searching::test::RunTests();
  return 0;
}

File: pattern_searching/kmp_pattern_searching.py
"""
KMP pattern searching algorithm
"""
import unittest


def kmp_pattern_searching(text: str, pattern: str) -> list[int]:
    if len(text) < 2 or len(pattern) < 2 or len(text) < len(pattern):
        return []

    lps: list[int] = [0] * len(pattern)
    i, j = 1, 0
    ret: list[int] = []

    lps[0] = 0
    while i < len(pattern) - 1:
        if pattern[i] == pattern[j]:
            j += 1
            lps[i] = j
            i += 1
        elif j > 0:
            j = lps[j - 1]
        else:
            lps[i] = 0
            i += 1

    i = j = 0
    while i < len(text) - 1:
        if text[i] == pattern[j]:
            i += 1
            j += 1

        if j == len(pattern) - 1:
            ret.append(i - j)
            j = lps[j - 1]
        elif i < len(text) - 1 and pattern[j] != text[i]:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1

    return ret


class TestKMPPatternSearching(unittest.TestCase):

    def test_handles_multiple_pattern_input(self):
        self.assertListEqual(kmp_pattern_searching('AABAACAADAABAAABAA', 'AABA'), [0, 9, 13])
        self.assertListEqual(kmp_pattern_searching('AAABSSSAAAOSI', 'FHJK'), [])
        self.assertListEqual(kmp_pattern_searching('AAACAAIOAAADAAIS', 'AAACA'), [0])

    def test_handles_invalid_input(self):
        self.assertListEqual(kmp_pattern_searching('', ''), [])
        self.assertListEqual(kmp_pattern_searching('ABC', ''), [])
        self.assertListEqual(kmp_pattern_searching('', 'DEF'), [])
        self.assertListEqual(kmp_pattern_searching('AA', 'DEF'), [])


if __name__ == '__main__':
    unittest.main()
